src/platform/android/CommissionableDataProviderImpl.cpp

```diff
@@ -0,0 +1,217 @@
+    std::vector<uint8_t> spake2pSalt(chip::Crypto::kSpake2p_Max_PBKDF_Salt_Length);
+
+    if (!havePaseSalt)
+    {
+        ChipLogProgress(Support, "LinuxCommissionableDataProvider didn't get a PASE salt, generating one.");
+        err = GeneratePaseSalt(spake2pSalt);
+        VerifyOrReturnLogError(err == CHIP_NO_ERROR, err);
+    }
+    else
+    {
+        chip::JniUtfString utfSpake2pSaltBase64(env, Spake2pSaltBase64);
+
+        size_t maxBase64Size = BASE64_ENCODED_LEN(chip::Crypto::kSpake2p_Max_PBKDF_Salt_Length);
+        VerifyOrReturnLogError(static_cast<unsigned>(utfSpake2pSaltBase64.size()) <= maxBase64Size, CHIP_ERROR_INVALID_ARGUMENT);
+
+        size_t decodedLen = chip::Base64Decode32(utfSpake2pSaltBase64.c_str(), utfSpake2pSaltBase64.size(),
+                                                 reinterpret_cast<uint8_t *>(spake2pSalt.data()));
+        VerifyOrReturnLogError(decodedLen >= chip::Crypto::kSpake2p_Min_PBKDF_Salt_Length &&
+                                   decodedLen <= chip::Crypto::kSpake2p_Max_PBKDF_Salt_Length,
+                               CHIP_ERROR_INVALID_ARGUMENT);
+        spake2pSalt.resize(decodedLen);
+    }
+
+    bool havePasscode = (setupPasscode > kMinSetupPasscode && setupPasscode < kMaxSetupPasscode);
+    Spake2pVerifier passcodeVerifier;
+    std::vector<uint8_t> serializedPasscodeVerifier(kSpake2p_VerifierSerialized_Length);
+    chip::MutableByteSpan saltSpan{ spake2pSalt.data(), spake2pSalt.size() };
+    if (havePasscode)
+    {
+        uint32_t u32SetupPasscode = static_cast<uint32_t>(setupPasscode);
+        err                       = passcodeVerifier.Generate(spake2pIterationCount, saltSpan, u32SetupPasscode);
+        VerifyOrReturnLogError(err == CHIP_NO_ERROR, err);
+
+        chip::MutableByteSpan verifierSpan{ serializedPasscodeVerifier.data(), serializedPasscodeVerifier.size() };
+        err = passcodeVerifier.Serialize(verifierSpan);
+        VerifyOrReturnLogError(err == CHIP_NO_ERROR, err);
+    }
+
+    // Make sure we actually have a verifier
+    VerifyOrReturnLogError(havePasscode || havePaseVerifier, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // If both passcode and external verifier were provided, validate they match, otherwise
+    // it's ambiguous.
+    if (havePasscode && havePaseVerifier)
+    {
+        VerifyOrReturnLogError(serializedPasscodeVerifier == serializedSpake2pVerifier, CHIP_ERROR_INVALID_ARGUMENT);
+        ChipLogProgress(Support, "Validated externally provided passcode matches the one generated from provided passcode.");
+    }
+
+    // External PASE verifier takes precedence when present (even though it is identical to passcode-based
+    // one when the latter is present).
+    std::vector<uint8_t> finalSerializedVerifier(kSpake2p_VerifierSerialized_Length);
+    if (havePaseVerifier)
+    {
+        finalSerializedVerifier = serializedSpake2pVerifier;
```

Comment:
why do we have a separate vector and do a copy? How about doing a `mSerializedPaseVerifier = std::move(serializedSpake2pVerifier)` here and a move of serializedPasscodeVerifier on the else branch?

Reply:
fixed. also added some comment on each step, but since each step is also simple enough ( 1 or 2 function call ), I did not add some functions, is that OK?